### PR TITLE
Add release and promote commands

### PR DESCRIPTION
**Pull request by robyoung**
This adds two new commands for the deployment pipeline; `release` and `promote`.

`release` builds a release based on the most recent pull request number, tag it in the repository and deploy it to the preview stage.

`promote` take a release (previously created by releasing to preview) and promotes it to either staging or production. When promoting to staging we need to copy the artefact from the development account into the production one. When promoting to production it ignores any provided release name and just promotes whatever is deployed into staging.

### CODE REVIEW

File: dmaws/build.py [hunk 1]
```diff
@@ -41,6 +44,50 @@ def get_current_ref(cwd):
...
+    return found_tag_name == tag_name
+
+
+def push_tag(cwd, tag_name, tag_message=None):
+    if tag_message is None:
+        tag_message = tag_name
+
+    run_git_cmd(['tag', '-a', tag_name, '-m', tag_message],
+                cwd).strip()
+    run_git_cmd(['push', 'origin', tag_name],
```
Comment (allait) [inline dmaws/build.py]: Won't this push the commits as well? This gets called for the development deploys too as far as I can tell, which might contain unpublished code

File: dmaws/commands/promote.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+def get_deploy(ctx, repository_path):
+    app = build.get_application_name(repository_path)
+    ctx.add_apps(app)
+
+    return StackPlan.from_ctx(ctx).get_deploy(repository_path)
+
+
+def promote_to_staging(ctx, repository_path, release_name, from_account):
+    if release_name is None:
+        raise StandardError("When promoting to staging a release name must be provided")
```
Comment (allait) [inline dmaws/commands/promote.py]: `ValueError` seems to fit quite well here

File: dmaws/deploy.py [hunk 3]
```diff
@@ -64,10 +90,20 @@ def upload_package(self, bucket_name, package_name, package_file):
 
         return bucket, key.key
 
+    def download_package(self, bucket_name, package_name):
+        bucket = self.conn.get_bucket(bucket_name)
+        key = bucket.get_key(package_name)
+        if not key:
+            raise StandardError('S3 key does not exist {}'.format(package_name))
+        package_file, package_path = tempfile.mkstemp()
```
Comment (allait) [inline dmaws/deploy.py]: `os.close(package_file)`

File: dmaws/build.py (hunk 1 above)
Comment (robyoung) [inline dmaws/build.py]: Discussed irl, gonna put this in the release task.

File: dmaws/commands/promote.py (hunk 2 above)
Comment (robyoung) [inline dmaws/commands/promote.py]: Good call.

File: dmaws/commands/release.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+
+@click.argument('repository_path', nargs=1, type=click.Path(exists=True))
+@click.option('--release-name')
+@click.option('--from-account')
+@cli_command('release', max_apps=0)
+def release_cmd(ctx, repository_path, release_name=None, from_account=None):
+    """Create a new application release and deploy.
+
+    This is more detail about the thing
```
Comment (allait) [inline dmaws/commands/release.py]: :smile:

**Pull request merged**